### PR TITLE
Expose modules to support creation of CDP DW Database Catalog and Virtual Warehouses

### PR DESCRIPTION
**Pull request by raju-saravanan**
Changes in this PR:
- Support private load balancers while creating a DW cluster.
- Support the creation of DW database catalog.
- Support the creation of DW virtual warehouse.

Changes in this PR are done in conjunction with the following PRs:
 - Deploy : https://github.com/cloudera-labs/cloudera-deploy/pull/46
 - Exe : https://github.com/cloudera-labs/cloudera.exe/pull/27
 - CDPY: https://github.com/cloudera-labs/cdpy/pull/23

### CODE REVIEW

File: plugins/modules/dw_dbc.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+DOCUMENTATION = r'''
+---
+module: dw_dbc
+short_description: Create CDP Data Warehouse Database Catalog
+description:
+    - Create CDP Database Catalog
+author:
+  - "Webster Mudge (@wmudge)"
+  - "Dan Chaffelson (@chaffelson)"
```
Comment (Chaffelson) [inline plugins/modules/dw_dbc.py:33]: I think you should be in the author list @raju-saravanan

File: plugins/modules/dw_dbc.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+        # Initialize return values
+        self.dbcs = []
+
+        # Execute logic process
+        self.process()
+
+    @CdpModule._Decorators.process_debug
+    def process(self):
+        self.name = self.cdpy.dw.create_dbc(cluster_id=self.cluster_id, name=self.name,
```
Comment (Chaffelson) [inline plugins/modules/dw_dbc.py:141]: What happens if it already exists?

File: plugins/modules/dw_vw.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+DOCUMENTATION = r'''
+---
+module: dw_vw
+short_description: Create CDP Data Warehouse Virtual Warehouse
+description:
+    - Create CDP Virtual Warehouse
+author:
+  - "Webster Mudge (@wmudge)"
+  - "Dan Chaffelson (@chaffelson)"
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:33]: Add authorship

File: plugins/modules/dw_vw.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        self.timeout = self._get_param('timeout')
+
+        # Initialize return values
+        self.vws = []
+
+        # Execute logic process
+        self.process()
+
+    @CdpModule._Decorators.process_debug
+    def process(self):
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:304]: These modules should be idempotent, and handle create or teardown

File: plugins/modules/dw_vw.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+            common_configs=dict(required=False, type='dict', aliases=['common_configs']),
+            application_configs=dict(required=False, type='dict', aliases=['application_configs']),
+            ldap_groups=dict(required=False, type='list', aliases=['ldap_groups']),
+            enable_sso=dict(required=False, type='bool', aliases=['enable_sso']),
+            tags=dict(required=False, type='dict', aliases=['tags']),
+            wait = dict(required=False, type='bool', default=True),
+            delay = dict(required=False, type='int', aliases=['polling_delay'], default=15),
+            timeout = dict(required=False, type='int', aliases=['polling_timeout'], default=3600)
+        ),
+        supports_check_mode=True
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:340]: You do not have check mode support in here, it should be added

File: plugins/modules/dw_dbc.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        self.dbcs.append(self.target)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=CdpModule.argument_spec(
+            cluster_id=dict(required=True, type='str', aliases=['cluster_id']),
+            name = dict(required=True, type='str', aliases=['name']),
+            load_demo_data=dict(required=False, type='bool', aliases=['load_demo_data']),
+            state=dict(required=False, type='str', choices=['present', 'absent'], default='present'),
```
Comment (Chaffelson) [inline plugins/modules/dw_dbc.py:160]: Please implement the absent state teardown controls, and make the module idempotent

**Review by Chaffelson — CHANGES_REQUESTED** [review]

File: plugins/modules/dw_vw.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+                        self.module.fail_json(msg="State %s is not valid for this module" % self.state)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=CdpModule.argument_spec(
+            cluster_id=dict(required=True, type='str', aliases=['cluster_id']),
+            dbc_id=dict(required=False, type='str', aliases=['dbc_id']),
+            vw_type = dict(required=False, type='str', aliases=['vw_type']),
+            name = dict(required=True, type='str', aliases=['name']),
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:405]: shouldn't be 'name' with an alias of 'name'

File: plugins/modules/dw_dbc.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+                            self.clusters.append(self.target)
+                        # Drop Done
+                elif self.state == 'present':
+                    # Begin Config check
+                    self.module.warn("DW Database Catalog already present and config validation is not implemented")
+                    if self.wait:
+                        self.target = self.cdpy.sdk.wait_for_state(
+                            describe_func=self.cdpy.dw.describe_dbc,
+                            params=dict(cluster_id=self.cluster_id,dbc_id=self.name),
+                            state='Running', delay=self.delay, timeout=self.timeout
```
Comment (Chaffelson) [inline plugins/modules/dw_dbc.py:185]: state should be self.cdpy.sdk.STARTED_STATES

File: plugins/modules/dw_vw.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+                                self.clusters.append(self.target)
+                            # Drop Done
+                    elif self.state == 'present':
+                        # Begin Config check
+                        self.module.warn("DW Virtual Warehouse already present and config validation is not implemented")
+                        if self.wait:
+                            self.target = self.cdpy.sdk.wait_for_state(
+                                describe_func=self.cdpy.dw.delete_vw,
+                                params=dict(cluster_id=self.cluster_id, vw_id=self.name),
+                                state='Running', delay=self.delay, timeout=self.timeout
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:364]: self.cdpy.sdk.STARTED_STATES

File: plugins/modules/dw_vw.py [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+                               dbc_id=self.dbc_id, vw_type=self.vw_type, name=self.name,
+                               template=self.template, autoscaling_min_cluster=self.autoscaling_min_cluster,
+                               autoscaling_max_cluster=self.autoscaling_max_cluster,
+                               common_configs=self.common_configs, application_configs=self.application_configs,
+                               ldap_groups=self.ldap_groups, enable_sso=self.enable_sso, tags=self.tags)
+                            if self.wait:
+                                self.target = self.cdpy.sdk.wait_for_state(
+                                    describe_func=self.cdpy.dw.describe_vw,
+                                    params=dict(cluster_id=self.cluster_id, vw_id=self.name),
+                                    state='Running', delay=self.delay, timeout=self.timeout
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:390]: self.cdpy.sdk.STARTED_STATES

**Review by Chaffelson — CHANGES_REQUESTED** [review]

File: plugins/modules/dw_vw.py [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+                        self.cdpy.sdk.sleep(3)  # Wait for consistency sync
+                        self.target = self.cdpy.dw.describe_vw(cluster_id=self.cluster_id, vw_id=self.target['id'])
+                        self.vws.append(self.target)
+                    # Drop Done
+            elif self.state == 'present':
+                # Begin Config check
+                self.module.warn("DW Virtual Warehouse already present and config validation is not implemented")
+                if self.wait:
+                    self.target = self.cdpy.sdk.wait_for_state(
+                        describe_func=self.cdpy.dw.delete_vw,
```
Comment (Chaffelson) [inline plugins/modules/dw_vw.py:366]: should be describe_vw not delete_vw!

**Review by Chaffelson — CHANGES_REQUESTED** [review]

File: plugins/modules/dw_vw.py (hunk 11 above)
Comment (raju-saravanan) [reply plugins/modules/dw_vw.py:366]: My bad

**Review by Chaffelson — APPROVED** [review]

File: plugins/modules/dw_vw.py [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+    required: False
+  autoscaling_min_cluster:
+    description: Minimum number of available nodes for Virtual Warehouse autoscaling.
+    type: int
+    required: False
+  autoscaling_max_cluster:
+    description: Maximum number of available nodes for Virtual Warehouse autoscaling.
+    type: int
+    required: False
+  config:
```
Comment (wmudge) [inline plugins/modules/dw_vw.py:71]: There is no `config` parameter group in the code -- `commonConfigs`, `applicationConfigs`, `ldapGroups`, and `enableSSO` are all "top level" module parameters in the code. Moreover, the parameters are lowercase with underscores.

File: plugins/modules/dw_vw.py [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+                    else:
+                        _ = self.cdpy.dw.delete_vw(cluster_id=self.cluster_id, vw_id=self.target['id'])
+                    if self.wait:
+                        self.cdpy.sdk.wait_for_state(
+                            describe_func=self.cdpy.dw.describe_vw,
+                            params=dict(cluster_id=self.cluster_id, vw_id=self.target['id']),
+                            field=None, delay=self.delay, timeout=self.timeout
+                        )
+                    else:
+                        self.cdpy.sdk.sleep(3)  # Wait for consistency sync
```
Comment (wmudge) [inline plugins/modules/dw_vw.py:357]: expose as option?

**Review by wmudge — CHANGES_REQUESTED** [review]
Looks good! I would like to clarify the module names and return values. In addition, the documentation sections need review, esp. the `dw_vw.py` module. The contents do not reflect the code, and I suspect the format of `applicationConfig`, for example,  needs updating, too.

File: plugins/modules/dw_dbc.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cloudera.cloud.plugins.module_utils.cdp_common import CdpModule
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: dw_dbc
```
[the remark this answers is not in the document]
Comment (raju-saravanan) [reply plugins/modules/dw_dbc.py:27]: going with `dw_database_catalog`

File: plugins/modules/dw_dbc.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+                    else:
+                        _ = self.cdpy.dw.delete_dbc(cluster_id=self.cluster_id, dbc_id=self.target['id'])
+                    if self.wait:
+                        self.cdpy.sdk.wait_for_state(
+                            describe_func=self.cdpy.dw.describe_dbc,
+                            params=dict(cluster_id=self.cluster_id, dbc_id=self.target['id']),
+                            field=None, delay=self.delay, timeout=self.timeout
+                        )
+                    else:
+                        self.cdpy.sdk.sleep(3)  # Wait for consistency sync
```
[the remark this answers is not in the document]
Comment (raju-saravanan) [reply plugins/modules/dw_dbc.py:182]: Will use already exposed `delay` option with sleep.

File: plugins/modules/dw_vw.py [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cloudera.cloud.plugins.module_utils.cdp_common import CdpModule
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: dw_vw
```
[the remark this answers is not in the document]
Comment (raju-saravanan) [reply plugins/modules/dw_vw.py:27]: Going with `dw_virtual_warehouse`

File: plugins/modules/dw_vw.py [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+    required: True
+  name:
+    description: Name of the Virtual Warehouse.
+    type: str
+    required: True
+  template:
+    description: Name of configuration template to use.
+    type: str
+    required: False
+  autoscaling_min_cluster:
```
[the remark this answers is not in the document]
Comment (raju-saravanan) [reply plugins/modules/dw_vw.py:63]: The cdpcli had 'cluster' as suffix for autoscaling parameters, I agree that it confusing. Will rename this to `autoscaling_min_nodes` etc.

File: plugins/modules/dw_vw.py (hunk 12 above)
Comment (raju-saravanan) [reply plugins/modules/dw_vw.py:71]: My bad, will fix this.

File: plugins/modules/dw_vw.py (hunk 13 above)
Comment (raju-saravanan) [reply plugins/modules/dw_vw.py:357]: Will use already exposed `delay` option with sleep.

**Comment by raju-saravanan** [comment]
@wmudge : Address your comments, can you take another look ?

File: plugins/modules/dw_virtual_warehouse.py [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+    required: False
+  autoscaling_max_nodes:
+    description: Maximum number of available nodes for Virtual Warehouse autoscaling.
+    type: int
+    required: False
+  common_configs: 
+    description: Configurations that are applied to every application in the service.
+    type: dict
+    required: False
+    contains:
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:75]: Apparently, `contains` has now been replaced by `suboptions` -- for example: https://github.com/ansible-collections/azure/blob/dev/plugins/modules/azure_rm_securitygroup.py#L49-L52

Full docs are here: https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block

File: plugins/modules/dw_virtual_warehouse.py [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+      configBlocks: List of ConfigBlocks for the application.
+        type: list
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:77]: Lists require `elements`, which would be `dict` in this case, IIRC

File: plugins/modules/dw_virtual_warehouse.py [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+            description: ID of the ConfigBlock. Unique within an ApplicationConfig.
+            type: str
+            required: False
+          format:
+            description: Format of ConfigBlock.
+            type: str
+            required: False
+          content:
+            description: Contents of a ConfigBlock.
+            type: obj
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:90]: `dict`

File: plugins/modules/dw_virtual_warehouse.py [hunk 21]
```diff
@@ -0,0 +1,433 @@
...
+            required: False
+            contains:
+              keyValues: 
+                description: Key-value type configurations. 
+                type: obj
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:95]: `dict`

File: plugins/modules/dw_virtual_warehouse.py [hunk 22]
```diff
@@ -0,0 +1,433 @@
...
+    required: False
+    default: 3600
+    aliases:
+      - polling_timeout  
+extends_documentation_fragment:
+  - cloudera.cloud.cdp_sdk_options
+  - cloudera.cloud.cdp_auth_options
+'''
+
+EXAMPLES = r'''
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:199]: Could you add an example or two of using the various nested configurations?

File: plugins/modules/dw_virtual_warehouse.py [hunk 23]
```diff
@@ -0,0 +1,433 @@
...
+                        self.target = self.cdpy.dw.describe_vw(cluster_id=self.cluster_id, vw_id=vw_id)
+                    self.virtual_warehouses.append(self.target)
+            else:
+                self.module.fail_json(msg="State %s is not valid for this module" % self.state)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=CdpModule.argument_spec(
+            id=dict(required=False, type='str', default=None),
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:402]: Only use `default` when necessary. The `_get_param()` should handle a non-existent parameter.

Comment (wmudge) [reply plugins/modules/dw_virtual_warehouse.py:402]: Same with `required` -- only needed if `True`

File: plugins/modules/dw_virtual_warehouse.py [hunk 24]
```diff
@@ -0,0 +1,433 @@
...
+            cluster_id=dict(required=True, type='str'),
+            dbc_id=dict(required=False, type='str', default=None),
+            type = dict(required=False, type='str', default=None),
+            name = dict(required=False, type='str', default=None),
+            template=dict(required=False, type='str', default=None),
+            autoscaling_min_nodes=dict(required=False, type='int', default=None),
+            autoscaling_max_nodes=dict(required=False, type='int', default=None),
+            common_configs=dict(required=False, type='dict', default=None),
+            application_configs=dict(required=False, type='dict', default=None),
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:411]: Both `common_configs` and `application_configs` need to be expanded to match documentation rules above.

File: plugins/modules/dw_virtual_warehouse.py [hunk 25]
```diff
@@ -0,0 +1,433 @@
...
+            state=dict(required=False, type='str', choices=['present', 'absent'], default='present'),
+            wait = dict(required=False, type='bool', default=True),
+            delay = dict(required=False, type='int', aliases=['polling_delay'], default=15),
+            timeout = dict(required=False, type='int', aliases=['polling_timeout'], default=3600)
+        ),
+        supports_check_mode=True
+    )
+
+    result = DwVw(module)
+    output = dict(changed=False, virtual_warehouses=result.virtual_warehouses)
```
Comment (wmudge) [inline plugins/modules/dw_virtual_warehouse.py:424]: `changed` is invalid -- this module can make changes.

**Review by wmudge — CHANGES_REQUESTED** [review]
The documentation and module specification sections need alignment and further detail.

**Comment by raju-saravanan** [comment]
@wmudge : I have expanded the `commonConfigs` in argument_spec, but `applicationConfig` have some dynamic keys in the uppermost level so, couldn't expand beyond 'dict'. Eg :
```
 "applicationConfigs": {
        "das-webapp": {
        "configBlocks": [
           {
               "id": "hive-kerberos-config",
               "format": "TEXT",
               "content": {
                "text": "\n[libdefaults]\n\trenew_lifetime = 7d"
              }
          }
        ] 
      }   
    }
```
Here "das-webapp" is dynamic key, we can have configs for other services as well.

**Comment by wmudge** [comment]
Superseded by #29

**Pull request closed without merging**